### PR TITLE
[FLINK-28701][runtime] Minimize the explosion range during failover for hybrid shuffle

### PR DESCRIPTION
**Pull request by reswqa**
## What is the purpose of the change

*In hybrid shuffle mode, there are currently two strategies to control spilling. For the full spilling strategy, the data is guaranteed to be persistent to the disk after task finished. When a failover occurs, if the upstream has been finished, the data should be recovered directly from the disk file without re-compute. For selective spilling strategy, the entire topology must be restarted.*

## Brief change log
  - *Split HYBRID result partition type to HYBRID_FULL and HYBRID_SELECTIVE.*
  - *RestartPipelinedRegionFailoverStrategy takes reConsumable into account when decide partition available.*
  - *Remove WIP prefix of ALL_EXCHANGE_HYBRID_FULL and ALL_EXCHANGE_HYBRID_SELECTIVE in BatchShuffleMode and expose them to documents.*

## Verifying this change

This change added new test methods in related test class.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no

### CODE REVIEW

File: flink-core/src/main/java/org/apache/flink/api/common/BatchShuffleMode.java
Comment (xintongsong) [inline flink-core/src/main/java/org/apache/flink/api/common/BatchShuffleMode.java:92]: Let's remove the WIP related things now. We can add another commit at the last of this PR.

**Review by xintongsong — APPROVED** [review]
This PR looks perfect to me. Well done, @reswqa. I have only one comment regarding removing the WIP status of this feature.

Let me know when the preceding PR is merged, and I'll merge this one.

**Review by xintongsong — APPROVED** [review]
Thanks @reswqa. LGTM. I've updated the configuration description a bit. I'll merge this once CI is passed the precedent PR is merged.

**Comment by xintongsong** [comment]
@flinkbot run azure

**Comment by reswqa** [comment]
@flinkbot run azure

**Pull request closed without merging**
